flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java

```diff
@@ -50,4 +54,17 @@
 	default boolean isUnknown() {
 		return false;
 	}
+
+	/**
+	 * Returns the location of the producing task executor if the partition occupies local resources there.
+	 *
+	 * <p>Indicates that this partition occupies local resources in the producing task executor. Such partition requires
+	 * that the task executor is running and being connected to be able to consume the produced data. This is mostly
+	 * relevant for the batch jobs and blocking result partitions which should outlive the producer lifetime and
+	 * be released externally: {@link ResultPartitionDeploymentDescriptor#isReleasedOnConsumption()} is {@code false}.
+	 * {@link ShuffleEnvironment#releasePartitions(Collection)} can be used to release such kind of partitions locally.
+	 *
+	 * @return the resource id of the producing task executor if the partition occupies local resources there
+	 */
+	Optional<ResourceID> hasLocalResources();
```

Comment:
if there are local resources, does this imply that there aren't _any_ external resources? (Basically, I want to double-check when we have to issue release calls to the ShuffleMaster)

Reply:
I would define it in a way that in general every shuffle service can potentially have some external resources if not then this is just an empty implementation. I would always call the external release in JM when the partition is not needed unless `releaseOnConsumption` is set in `ResultPartitionDeploymentDescriptor`.